### PR TITLE
Simplify knight and ranged piece movements, remove null return values

### PR DESCRIPTION
**Pull request by ddugovic**
I am curious whether it is possible to speed up builds and/or to simplify code.

### CODE REVIEW

File: playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/models/position/ShogiBoardStateImpl.java [hunk 1]
```diff
@@ -24,12 +26,12 @@ private ShogiBoardStateImpl(final int width, final int height, final Piece[][] b
     }
 
     @Override
-    public Piece getPieceAt(final int column, final int row) {
-        return board[column - 1][row - 1];
+    public Optional<Piece> getPieceAt(final int column, final int row) {
```
Comment (Tellmarch) [inline playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/models/position/ShogiBoardStateImpl.java:29]: There is some performance cost to returning an optional on such a basic operation: every time that we look at a non-empty square of the board, it will create an object (also resulting in more frequent garbage collection).
Since we are not writing a Shogi engine with this library I think it's fine (I'm not asking to revert the change) - just something to keep in mind

File: playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/rules/movements/BishopMovement.java [hunk 2]
```diff
@@ -20,8 +20,7 @@
 
     @Override
     public boolean isMoveDxDyValid(final ShogiBoardState boardState, final Square from, final Square to) {
-        // TODO write more efficient method?
-        return getPossibleMoves(boardState, from).contains(to);
+        return PieceMovementsUtils.isAlongDirection(boardState, from, to);
```
[the remark this answers is not in the document]
Comment (ddugovic) [reply playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/rules/movements/BishopMovement.java:23]: Ah, yes I should add a unit test and correct this.

File: playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/models/position/ShogiBoardStateImpl.java (hunk 1 above)
Comment (ddugovic) [reply playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/models/position/ShogiBoardStateImpl.java:29]: Thanks, yes I did consider that there is some slight overhead and some slight code complexity, although protecting against a potential subtle NullPointerException seems worth it, at least for long-lived references or values.

**Comment by Tellmarch** [comment]
Thanks!
